cmd/tlssocksproxy/tlssocksproxy.go

```diff
@@ -153,19 +164,52 @@ func recoverAndLogPanic() {
 		case error:
 			err = x
 		default:
-			// Fallback err (per specs, error strings should be lowercase w/o punctuation
-			err = errors.New("unknown panic")
+			err = fmt.Errorf("unknown panic of type: %T", r)
 		}
 		logger.Error("Panic occurred in serve thread", zap.Error(err))
 	}
 }
 
-func runPrometheusHandler(_ context.Context, address string) {
+func runPrometheusHandler(ctx context.Context, logger *zap.Logger, address string) {
 	h := http.NewServeMux()
 	h.Handle("/metrics", promhttp.Handler())
-	logger.Fatal("Failed to start prometheus handler", zap.Error(http.ListenAndServe(address, h)))
+	server := &http.Server{Addr: address, Handler: h}
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			logger.Fatal("Failed to start prometheus handler", zap.Error(err))
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
+		logger.Info("Shutdown prometheus handler in progress")
+		if err := server.Shutdown(ctx); err != nil && err != context.Canceled {
+			logger.Fatal("Failed to Shutdown prometheus handler", zap.Error(err))
+		}
+	}
 }
 
 func silentClose(closer io.Closer) {
 	_ = closer.Close()
 }
+
+func ctxCancelOnOsSignal(logger *zap.Logger) context.Context {
```

Comment:
Shouldn't this be on all three  services? Like a common thing?

Reply:
fixed